### PR TITLE
Aggregations Refactor: Refactor Bucket Selector Aggregation

### PR DESCRIPTION
**Pull request by colings86**

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java [hunk 1]
```diff
@@ -136,20 +141,83 @@ protected void doReadFrom(StreamInput in) throws IOException {
...
         protected PipelineAggregator createInternal(Map<String, Object> metaData) throws IOException {
             return new BucketSelectorPipelineAggregator(name, bucketsPathsMap, script, gapPolicy, metaData);
         }
+
+        @Override
+        protected XContentBuilder internalXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.field(BucketScriptParser.BUCKETS_PATH.getPreferredName(), bucketsPathsMap);
+            builder.field(ScriptField.SCRIPT.getPreferredName(), script);
+            builder.field(BucketScriptParser.GAP_POLICY.getPreferredName(), gapPolicy.getName());
+            return builder;
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java]: Is this new or code that was moved? I was looking for where this came from but couldn't find anything.

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java [hunk 2]
```diff
@@ -136,20 +141,83 @@ protected void doReadFrom(StreamInput in) throws IOException {
...
+        }
+
+        @Override
+        protected boolean overrideBucketsPath() {
+            return true;
+        }
+
+        @Override
+        protected PipelineAggregatorFactory doReadFrom(String name, String[] bucketsPaths, StreamInput in) throws IOException {
+            Map<String, String> bucketsPathsMap = new HashMap<String, String>();
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java]: I was wondering wherer the bucketsPaths passed get used here, but they might only be necessary in other impls of PipelineAggregatorFactory I guess. I also found the order in which the super class reads something from the stream, then calls this method to create the actual implementation and then reads some more a bit hard to follow and a bit tricky. Maybe there are ways of simplifying this, but unfortunaltely I don't have a good suggestion at this point. Maybe the whole readFrom/doReadFrom separation in the abstract class and the implementations could be merged, which would mean some duplication in code but would be easier to understand and maintain? Just a suggestion though.

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java (hunk 1 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java]: This is new but reflects the same output as BucketSelectorBuilder provides

**Comment by cbuescher** [comment]
Looks good to me, left one question and a general remark concerning the way reading/writing objects is spread across abstract classes and implementations, but this is something than can be revisited later.

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java (hunk 2 above)
Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java]: bucketsPaths doesn't get used here as this implementation overrides bucketsPaths and uses its own parameter (a map instead of an array) so each bucketPath can be named and then referenced by that name in the script. 

In terms of the superclass reading stuff and then passing to the subclass I think it needs to be this as the superclass needs to read the stuff that it requires (mandatory options) to be passed into its constructor but then has to pass this stuff to the subclass since it doesn't know how to construct the instance of the subclass. It then has to set the optional parameters on the returned instance so they are set properly. I wouldn't want each subclass to have to separately read and set all the options which are generic to all pipeline aggregator factories.

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/PipelineAggregatorFactory.java [hunk 3]
```diff
@@ -158,6 +158,10 @@ public final XContentBuilder toXContent(XContentBuilder builder, Params params)
         return builder.endObject();
     }
 
+    protected boolean overrideBucketsPath() {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/PipelineAggregatorFactory.java]: Can you add javadocs?

Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/PipelineAggregatorFactory.java]: Will do. This is actually the same change as made in https://github.com/elastic/elasticsearch/pull/15014 where I have already added JavaDocs. I will make sure that one is merged first and the JavaDoc is retained

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java (hunk 2 above)
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java]: I'm not very happy with the buckets_path parsing/xcontent either. Would it be possible for PipelineAggregatorFactory to not care about bucketsPath at all and then have a sub-class that handles the case when bucketsPath is an array so that most classes don't have to reimplement the same logic. Most pipeline agg factories would extend this class while BucketSelectorPipelineAggregator would extend PipelineAggregatorFactory directly?

Comment (colings86) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java]: PipelineAggregatorFactory has to care about bucketsPaths because it need to use them in order to work out the dependency between the pipeline aggregation and the other aggregations to know what order to run them in. I guess we could have a getBucketsPathsArray() method that returns String[] and the PipelineAggregatorFactory could use that for its dependency resolution stuff? That would mean we could have the subclass that you talked about. But we are still going to have the name read from the stream before calling doReadFrom() and then other stuff read by PipelineAggregatorFactory.readFrom() after the doReadFrom() has returned.

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/having/BucketSelectorPipelineAggregator.java]: OK sorry, I had missed the dependency resolution issue. Then I suggest to keep your current solution for potentially revisit when all aggs are refactored and we have a better picture.

File: core/src/main/java/org/elasticsearch/search/aggregations/pipeline/PipelineAggregatorFactory.java (hunk 3 above)
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/pipeline/PipelineAggregatorFactory.java]: ++

**Comment by jpountz** [comment]
LGTM
